watchmal/dataset/cnn_mpmt/cnn_mpmt_dataset.py

```diff
@@ -45,13 +56,53 @@ def get_data(self, hit_pmts, hit_charges, hit_times):
         # collapse arrays if desired
         if self.collapse_arrays:
             data = np.expand_dims(np.sum(data, 0), 0)
+        
         processed_data = from_numpy(data)
 
         if self.transforms is not None:
             selection = np.random.choice(2, self.n_transforms)
             for i, transform in enumerate(self.transforms):
                 if selection[i]:
                     processed_data = transform(processed_data)
-        
 
         return processed_data
+
+    def retrieve_event_data(self, item):
+        """
+        Returns event data from dataset associated with a specific index
+
+        Args:
+            item                    ... index of event
+        
+        Returns:
+            hit_pmts                ... array of ids of hit pmts
+            pmt_charge_data         ... array of charge of hits
+            pmt_time_data           ... array of times of hits
+
+        """
+        start = self.event_hits_index[item]
+        stop = self.event_hits_index[item + 1]
+
+        hit_pmts = self.hit_pmt[start:stop].astype(np.int16)
+        hit_charges = self.charge[start:stop]
+        hit_times = self.time[start:stop]
+
+        hit_mpmts = hit_pmts // pmts_per_mpmt
+        hit_pmt_in_modules = hit_pmts % pmts_per_mpmt
+
+        hit_rows = self.mpmt_positions[hit_mpmts, 0]
+        hit_cols = self.mpmt_positions[hit_mpmts, 1]
+
+        # construct charge data with barrel array indexing to match endcaps in xyz ordering
+        charge_data = np.zeros(self.data_size)
+        charge_data[hit_pmt_in_modules, hit_rows, hit_cols] = hit_charges
+        charge_data[:, 12:28, :] = charge_data[barrel_map_array_idxs, 12:28, :]
+        pmt_charge_data = charge_data[hit_pmt_in_modules, hit_rows, hit_cols].flatten()
+
+        # construct time data with barrel array indexing to match endcaps in xyz ordering
+        time_data = np.zeros(self.data_size)
+        time_data[hit_pmt_in_modules, hit_rows, hit_cols] = hit_charges
+        time_data[:, 12:28, :] = time_data[barrel_map_array_idxs, 12:28, :]
+        pmt_time_data = charge_data[hit_pmt_in_modules, hit_rows, hit_cols].flatten()
+
+        return hit_pmts, pmt_charge_data, pmt_time_data
```

Comment:
This code seems to be largely duplicated from the `H5Dataset`'s `__getitem__` and this file's `get_data`. Is there a way we can avoid this by pulling it out into separate functions? Perhaps a better way of arranging it would be to remove the `__getitem__` form the abstract `H5Dataset` and move that code (removing the `get_data` call) into some other function (which could be called e.g. `get_data` like the subclass currently has. This `retrieve_event_data` method can just call that. Then the subclass defines a `__get_item__`, which simply calls the `retrieve_event_data` and then does the transformations. Will that work without any code duplication? If that's not clear, I can try to make these modifications myself later.

Reply:
Looks good now